src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,959 @@
+     * Then the method will check if the packet is known to the registry, and if it is not, null is returned.
+     * If the packet is known to the registry, the method will return a reader that will invoke the registered replyHandler.
+     * </p>
+     *
+     * @param id         The id of the payload.
+     * @param context    The context of the channel.
+     * @param protocol   The protocol of the connection.
+     * @param knownTypes The known types of the connection.
+     * @return A reader for the payload, or null if the payload should be discarded.
+     */
+    public FriendlyByteBuf.Reader<? extends CustomPacketPayload> getReader(ResourceLocation id, ChannelHandlerContext context, ConnectionProtocol protocol, Map<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>> knownTypes) {
+        //Vanilla custom packet, let it deal with it.
+        if (knownTypes.containsKey(id)) {
+            return knownTypes.get(id);
+        }
+
+        //These are our own custom modded packets which can be sent before a payload setup is negotiated.
+        //Special case them
+        if (id.equals(ModdedNetworkQueryPayload.ID)) {
+            return ModdedNetworkQueryPayload.READER;
+        }
+
+        if (id.equals(ModdedNetworkPayload.ID)) {
+            return ModdedNetworkPayload.READER;
+        }
+
+        if (id.equals(ModdedNetworkSetupFailedPayload.ID)) {
+            return ModdedNetworkSetupFailedPayload.READER;
+        }
+
+        //Check the network setup.
+        final NetworkPayloadSetup payloadSetup = context.channel().attr(ATTRIBUTE_PAYLOAD_SETUP).get();
+        final PacketFlow flow = context.channel().attr(ATTRIBUTE_FLOW).get();
+        if (payloadSetup == null || flow == null) {
+            //Error: Bail.
+            LOGGER.warn("Received a modded custom payload packet {} that has not negotiated with the server. Not parsing packet.", id);
+            return null;
+        }
+
+        //Now ask the protocol what kind of payload is being sent and get the channel for it.
+        if (protocol.isPlay()) {
+            final NetworkChannel channel = payloadSetup.play().get(id);
+
+            //Validate that everything is okey and then return a reader.
+            if (channel == null) {
+                LOGGER.warn("Received a modded custom payload packet {} with an unknown or not accepted channel. Not parsing packet.", id);
+                return null;
+            }
+
+            final PlayRegistration<?> registration = knownPlayRegistrations.get(channel.id());
+            if (registration == null) {
+                LOGGER.error("Received a modded custom payload packet {} with an unknown or not accepted channel. Not parsing packet.", channel.id());
+                throw new IllegalStateException("A client send a packet with an unknown or not accepted channel, while negotiation succeeded. Somebody changed the channels known to NeoForge!");
```

Comment:
`sent` not `send`. Also, that assumption in the exception sounds.. wrong and should be removed as it jumps to conclusions.

Reply:
See above